src/lib/Inference.hs

```diff
@@ -1531,6 +1532,24 @@ binderNestAsPiNest arr = \case
   Empty             -> Empty
   Nest (b:>ty) rest -> Nest (PiBinder b ty arr) $ binderNestAsPiNest arr rest
 
+inferRoleBinderNest :: Fallible m => Nest Binder n l -> m (Nest RoleBinder n l)
+inferRoleBinderNest Empty = return Empty
+inferRoleBinderNest (Nest (b:>ty) bs) =
+  Nest (RoleBinder b ty (typeToRole ty)) <$> inferRoleBinderNest bs
+{-# INLINE inferRoleBinderNest #-}
+
+inferRolePiBinderNest :: Fallible m => Nest PiBinder n l -> m (Nest RolePiBinder n l)
+inferRolePiBinderNest Empty = return Empty
+inferRolePiBinderNest (Nest (PiBinder b ty arr) bs) =
+  Nest (RolePiBinder b ty arr (typeToRole ty)) <$> inferRolePiBinderNest bs
+{-# INLINE inferRolePiBinderNest #-}
+
+typeToRole :: Type n -> ParamRole
+typeToRole ty = case ty of
+  TyKind -> TypeParam
+  DictTy _ -> DictParam
+  _ -> DataParam -- TODO(dougalm): check that it's actually "data" and throw a useful error if not
```

Comment:
Since this is entirely type-driven, why do we have to keep those role annotations in binders? We could just resynthesize them whenever needed, right? It would save us the new `Role*` types and a bunch of casts between PiBinders and those

Reply:
I'm torn on this one. The original intention was to treat role inference separate from rule use. (Just as we like to do with type inference.) That way we can change the inference rules if we like, and we guarantee that we only assign a single role, and we can allow explicit user-annotated roles or whatever. But the implementation turned out to make the role-as-a-function-of-type assumption more explicitly than I'd expected. I'm still inclined to keep it as it is, mainly because it separates "what should the role be" from "how should this role behave".

Reply:
Right, I guess a good test for this might be to have a dict that's passed in as an explicit argument somewhere. We do that for `Accum` after all. Even though the type of the dict would indicate that it should have a dict param role, it's an explicit argument, so we should not assume coherence.

So I buy the annotation argument I think. But perhaps the roles should be derived from the arrows _and_ types, not just types.